File: alchemist/alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/implementations/movestrategies/speed/ConstantSpeed.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.implementations.movestrategies.speed;

import it.unibo.alchemist.model.interfaces.Position;
import it.unibo.alchemist.model.interfaces.Reaction;
import it.unibo.alchemist.model.interfaces.movestrategies.SpeedSelectionStrategy;

/**
 * This strategy makes the node move at an average constant speed, which is
 * influenced by the {@link it.unibo.alchemist.model.interfaces.TimeDistribution} of the {@link Reaction} hosting
 * this {@link it.unibo.alchemist.model.interfaces.Action}. This action tries to normalize on the {@link Reaction}
 * rate, but if the {@link it.unibo.alchemist.model.interfaces.TimeDistribution} has a high variance, the movements
 * on the map will inherit this tract.
 *
 * @param <P> Position type
 * 
 */
public final class ConstantSpeed<P extends Position<P>> implements SpeedSelectionStrategy<P> {

    private static final long serialVersionUID = 1746429998480123049L;
    private final double speed;
    private final Reaction<?> reaction;

    /**
     * @param reaction
     *            the reaction
     * @param speed
     *            the speed, in meters/second
     */
    public ConstantSpeed(final Reaction<?> reaction, final double speed) {
        if (speed < 0) {
            throw new IllegalArgumentException("Speed must be positive");
        }
        this.speed = speed;
        this.reaction = reaction;
    }

    @Override
    public double getNodeMovementLength(final P target) {
        return speed / reaction.getRate();
    }

}

File: alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/movestrategies/speed/ConstantSpeed.kt
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.movestrategies.speed

import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.Position
import it.unibo.alchemist.model.Reaction
import it.unibo.alchemist.model.movestrategies.SpeedSelectionStrategy
import java.io.Serial

/**
 * This strategy makes the node move at an average constant speed, which is
 * influenced by the [it.unibo.alchemist.model.TimeDistribution] of the [Reaction] hosting
 * this [it.unibo.alchemist.model.Action]. This action tries to normalize on the [Reaction]
 * rate, but if the [it.unibo.alchemist.model.TimeDistribution] has a high variance, the movements
 * on the map will inherit this tract.
 *
 * @param <T> Concentration type
 * @param <P> Position type
 */
class ConstantSpeed<T, P : Position<P>>(
    private val reaction: Reaction<*>,
    private val speed: Double,
) : SpeedSelectionStrategy<T, P> {
    /**
     * @param reaction the reaction
     * @param speed the speed, in meters/second
     */
    init {
        require(speed >= 0) { "Speed must be positive or zero in $reaction. Provided: $speed" }
    }

    override fun getNodeMovementLength(target: P?): Double = speed / reaction.rate

    override fun cloneIfNeeded(
        destination: Node<T>,
        reaction: Reaction<T>,
    ): ConstantSpeed<T, P> = ConstantSpeed(reaction, speed)

    private companion object {
        @Serial
        private const val serialVersionUID = 1746429998480123049L
    }
}
